text/0082-entity-tags.md

```diff
@@ -0,0 +1,251 @@
+# Entity tags with dedicated syntax and semantics
+
+## Related issues and PRs
+
+* Reference Issues: [#305](https://github.com/cedar-policy/cedar/issues/305)
+* Supersedes [RFC 68](https://github.com/cedar-policy/rfcs/pull/68)
+* Implementation PR(s): (leave this empty)
+
+## Timeline
+
+* Start Date: 2024-09-11
+* Date Entered FCP: 2024-09-11
+* Date Accepted:
+* Date Landed:
+
+## Summary
+
+This RFC proposes to extend the Cedar language, type system, and symbolic analysis to include full-featured *entity tags* for entity types. Tags are a mechanism used by cloud services to attach key-value pairs to resources. Cedar will allow them to be attached to any entities (not just resources).
+
+For evaluation purposes, entity tags are accessed using the methods `expr1.hasTag(expr2)` and `expr1.getTag(expr2)`, where `expr1` evaluates to an entity and `expr2` to a string.  These are the only ways to access the tags on an entity.  The policy syntax, semantics, type system, and symbolic analysis are extended to support these two methods.  The schema syntax and the entity JSON format are extended to support defining entity tags and their types.
+
+Entity tags strictly increase the expressiveness of Cedar because they allow *computed keys*, unlike records and entities. Specifically, tag keys need not be literals, so it is possible to write `resource.hasTag(context.tag)` and `resource.getTag(context.tag)`.
+
+This proposal is backward-compatible in the sense that existing polices, schemas, entities, etc., do not need to change to accommodate the addition of tags to the language.
+
+This RFC has gone through several revisions, with different revisions proposing alternative designs. The alternatives are discussed at the end of this document.
+
+### Basic example
+
+Here is a schema defining two entities, each of which contains entity tags.
+
+```
+entity User = {
+  jobLevel: Long,
+} tags Set<String>;
+entity Document = {
+  owner: User,
+} tags Set<String>;
+```
```

Comment:
One concern I thought of with this syntax: If you write
```
entity User = {
    jobLevel: Long,
} tags { foo: String };
```
this is valid but potentially very confusing.  It might look a lot like the entity has a required tag `foo` of type `String`, but that is not at all what that means in the current proposal.  Instead, the entity has (zero or more) tags, with any names, and the value for all of them is a record with attribute `foo` of type `String`.

I'm not sure if there is a clearer syntax alternative.  Ideas?

Reply:
I personally think it's Ok. Not having curly braces is clear, e.g., `entity { ... } tags Set<String>`. Having them, they evoke that the tags are a record. If we wrote `entity User { ... } tags { foo: String, bar: String }` what would it mean otherwise? Adding extra delimiters seems unnecessary (which could not be curly braces).